### ISSUE TITLE
Refinement to Iterator interface

### ISSUE THREAD

**Issue by brightredchilli** [keras-team/keras#9131]
Hi all, I'm looking through the source of keras.preprocessing.image.Iterator, and specifically, the requirement that Iterators implement `_get_batches_of_transformed_samples()` I'm wondering if it might be good to move in a direction where Iterators have a default image_generator=None, and instead require `_get_batches_of_samples()`. In that case, DirectoryIterators can be instantiated without an ImageDataGenerator, and won't crash on this [line](https://github.com/keras-team/keras/blob/950e5d063320a72ca61f2082c154a65a48766239/keras/preprocessing/image.py#L1150).

```python
# current 
def _get_batches_of_transformed_samples(self, index_array):
    ...
 x = self.image_data_generator.random_transform(x)

# proposal
def _get_batches_of_samples(self, index_array):
    ...
    if self.image_data_generator:
        x = self.image_data_generator.random_transform(x)
```

The use case that i'm running into is that I want to capture some bottleneck features using a generator. Since generators can't be reset, I need to create one generator when running all my data through my model, and another specifically to capture the output classes. Looking through the DirectoryIterator, it nicely sorts of the dirs and filenames, so I'm convinced that on separate runs(if `shuffle=False`), the order of the labels remains the same. It would just be nice if I didn't have to run the image transformation when getting the labels(it takes a longer time!).

However, I could definitely see someone wanting to use DirectoryIterator without ImageDataGenerator.
